### DISCUSSION TITLE
getStaticPaths with fallback: false no longer returns 404 in v12 or 13

### DISCUSSION THREAD

**Discussion by brianconcannon** [vercel/next.js#44665] [Help]
Upgrading from next 11.1.2 to 12 or 13 results in changed behavior for dynamic URLs that used to return 404s. I have the following: /pages/[searchTerm].js with thousands of paths that are generated during build time with fallback: false returned by getStaticPaths(). getStaticProps() gets the data for those pages at build time, but doesn't run in production.

Previously (v11), when a user searches for a page that wasn't prebuilt (i.e. doesn't exist) it would return a 404 and redirect to my /pages/404.js page. In v12 and beyond it no longer produces a 404 (in production), instead it produces a 304 and then tries to render that page, but without any data. Note: I currently don't have it trying to dynamically load data because it is all done during build.

The other factor making this difficult to troubleshoot is that it works differently in a development environment (npm run dev) than it does when deployed with next run export (my site is served statically from s3). I get the 404s in the dev environment and the redirect to 404.js.

I've scoured the docs to see if any changes were made to dynamic URLs with fallback: false and I can't find anything other than that it should work like it used to, but for some reason it doesn't. But a lot has changed since I last upgraded, so any help would be appreciated. Thanks in advance!

**Comment by icyJoseph** [comment]
Hi,

I've been able to reproduce this.

## Reproduction

I think it is **a serious bug**. In **next@12.3.4**, going to a page with fallback false returns 404 and shows the 404 page. This is correct.

**In next@13.1.1, I see a 404 but immediately the page tries to render.**

To reproduce it, _**initiate middleware**_ and have a page like: `pages/posts/[page].tsx`:

```tsx
import type { GetStaticPaths, GetStaticProps, NextPage } from "next";

export const PostsPage: NextPage<{
  total: number;
  page: string;
}> = (props) => {
  const { total, page } = props;

  return (
    <section>
      <h1>Post</h1>

      <p>
        Post page: {page} from a total: {total}
      </p>
    </section>
  );
};

/* get this from your CMS */
const paths = Array.from({ length: 4 }, (_, i) => ({
  params: { page: `${i}` },
}));

export const getStaticPaths: GetStaticPaths<{ page: string }> = () => {
  return {
    paths,
    fallback: false,
  };
};

export const getStaticProps: GetStaticProps = async (ctx) => {
  const page = ctx.params?.page;

  if (typeof page === "undefined") return { notFound: true };
  if (Array.isArray(page)) return { notFound: true };

  return {
    props: {
      page,
      total: paths.length,
    },
  };
};

export default PostsPage;
```

Now build and start, and go to something like `posts/100`, it'll flash 404, and even return 404 to the request, but the client side runtime tries to render the UI anyway. This is the bug.

![Screenshot 2023-01-07 at 16 52 21](https://user-images.githubusercontent.com/21013447/211159221-fc6cd617-9366-4c0d-a113-494d464c3e2f.png)

## Without middleware

Now if we **remove the middleware**, and build + start:

![Screenshot 2023-01-07 at 16 53 21](https://user-images.githubusercontent.com/21013447/211159309-6609b2c9-dc6c-4b97-88b3-37c2dac7ddb9.png)


Follow up questions:

- When you say that next 12 is broken, which version exactly? Did you upgrade from 11 to 13, saw the issue and downgraded to 12? did you delete `.next` and `node_modules` when doing that? I want to pin-point a poisoned version 12.
- Do you think you can open a bug report?

## What to do now

I think some changes were made to allow https://nextjs.org/blog/next-13-1#nextjs-advanced-middleware, but then we should see the errors in 12 too, because of some those were sneaked in without [release notes](https://github.com/vercel/next.js/discussions/43261), but it might be a false lead. 

I'll try to clear up time to try and see why the router is trying to do this. I'll also investigate what's up in dev mode.

## Edit for dev mode

In Dev mode I cannot reproduce this. I am now afraid that we are looking at two bugs. 

## Possible introduction

`next@13.0.7-canary.5` And that means that the commit that introduced it, what I found anyway, is here: https://github.com/vercel/next.js/compare/v13.0.7-canary.4...v13.0.7-canary.5 mmmm

Perhaps: https://github.com/vercel/next.js/commit/cba7070c0856c770eae685dd3e5a92c8f2e6e366 ?

**Comment by brianconcannon** [comment]
Yes, I tried to upgrade to 13.1.1 and then backtracked to 12. I’ve tried 12.3.4 so far and this is what I see. In dev, I do get the 404 and pages/404.js runs as expected, but with a new router error:

next-dev.js?3515:20 Error: Failed to load static props
    at eval (webpack-internal:///./node_modules/next/dist/shared/lib/router/router.js:556:23)

However, there's a difference in a production build for 12.3.4. It does not throw a 404, it responds with a 200 and doesn't redirect to pages/404.js and continues to try to render that missing page.

I'll backtrack further into older 12 versions and see if I can pinpoint the diversion from the behavior I was used to in 11.1.2.

**Comment by brianconcannon** [comment]
Ok, one more update. I went back to 11.1.2 where I've been for a while and even then the router provides a 404 in dev for a pages/[foo].js that doesn't exist and a 304 in prod, but it does redirect to pages/404.js without trying to render the dynamic page.

I worked my way back to 12.0.0 and that is where things are different. Dev works as expected, but a production build doesn't redirect to pages/404.js and tries to render the dynamic page without props because getStaticProps never ran for that missing page.

Let me know if you still want me to convert this to a bug report (or open a new one).

**Reply by icyJoseph** [reply]
Have you been clearing the browser redirect cache? You have to open dev tools and ctrl/cmd+shift+r - not sure 304's are stored though. Perhaps it is better to use incognito mode when testing this stuff.

That's so strange, I cannot see the same for 12.0.0, what could the issue be. At this point, I'd say if you can create a next app with create-next-app and reproduce this, it'd be much better.

I can for sure see an issue with 13.1.1 - which started at 13.0.7-canary.5 

Could please try to make a reproduction repository?

**Comment by brianconcannon** [comment]
I created a new next app and tried to reproduce this error with 13.1.1. I found that the apps works fine with default settings, but if you add trailingSlash: true to next.config.js, that's when it produces the unexpected behavior where it doesn't route to the 404 page, but instead tries to render the dynamic page.

**Reply by brianconcannon** [reply]
This only happens on the exported app in production. My site is served statically from s3/cloudfront and the routing bug only appears when the exported app is served from there. There might be a way to reproduce it locally, but I haven't been able to yet. I prefer to use trailingSlash: true in my app so that each page gets a subfolder with an index.html and my URLS don't need to reference the .html extension. 

I still think this is a bug, but perhaps not as big as first thought. It's simply a routing issue for missing dynamic pages with fallback: false and trailingSlash: true for statically exported sites.

**Reply by icyJoseph** [reply]
Aha that explains a lot though! So I did find yet another issue though. 

Exported apps expect a lot from the serving engine and trailing slash handling is always a suspect. I'll re-read this thread later today though.
